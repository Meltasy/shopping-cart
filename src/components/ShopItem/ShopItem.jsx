import { useState } from 'react'
import styled from 'styled-components'
import DetailItem from '../DetailItem/DetailItem'
import ItemForm from '../ItemForm/ItemForm'

const Wrapper = styled.div`
  width: 300px;
  height: 550px;
  padding: 10px;
  color: var(--text-color);
  background-color: white;
  border: 2px solid var(--primary-color);
  border-radius: 10px;
  justify-items: center;
  align-self: center;
  align-content: end;
`

const StyledImage = styled.img`
  max-width: 225px;
  max-height: 225px;
  object-fit: cover;
  border-radius: 10px;
`

const StyledButton = styled.button`
  font-family: inherit;
  font-size: inherit;
  color: var(--background-color);
  background-color: var(--primary-color-dark);
  padding: 5px 10px;
  border: 2px solid var(--primary-color);
  border-radius: 5px;
  margin: 5px 0;
`

const ShopItem = ({ itemId, imageUrl, title, price, description, handleAdd }) => {
  const [showDetail, setShowDetail] = useState(false)

  return (
    <Wrapper key={itemId}>
      <StyledImage src={imageUrl} alt=' ' />
      <h3>{title} â‚¬{price ? price.toFixed(2) : '0.00'}</h3>
      <StyledButton onClick={() => setShowDetail(true)}>Show Detail</StyledButton>
      <DetailItem
        showDetail={showDetail}
        itemId={itemId}
        title={title}
        price={price}
        imageUrl={imageUrl}
        description={description}
      >
        <StyledButton onClick={() => setShowDetail(false)}>Back to Shop</StyledButton>
      </DetailItem>
      <ItemForm
        itemId={itemId} price={price} onSubmit={handleAdd} />
    </Wrapper>
  )
}

export default ShopItem
