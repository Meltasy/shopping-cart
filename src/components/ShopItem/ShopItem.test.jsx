import { vi, describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import ShopItem from './ShopItem'

describe('Shop item component', () => {
  it('renders shop product items', () => {
    render(<ShopItem imageUrl='fakeProduct.jpg' title='Fake product' price={35.50} />)
    const { container } = render(<ShopItem />)
    expect(container).toMatchSnapshot()
  })

  it('renders correct shop item details when found', () => {
    render(<ShopItem imageUrl='fakeProduct.jpg' title='Fake product' price={35.50} />)
    screen.debug()
    expect(screen.getAllByRole('img', { hidden: true })[0]).toHaveAttribute('src', 'fakeProduct.jpg')
    expect(screen.getByRole('heading', { hidden: true, level: 3 })).toHaveTextContent('Fake product €35.50')
  })

  it('does not render shop item details when not found', () => {
    render(<ShopItem imageUrl='fakeProduct.jpg' title='Fake product' price={57.99} />)
    expect(screen.queryByText('€57.99')).not.toBeInTheDocument()
  })
})

describe('Shop item component detail buttons', () => {
  it('renders all two buttons', () => {
    render(<ShopItem />)
    expect(screen.getAllByRole('button').length).toBe(4)
  })

  it('renders all buttons for the DetailItem dialog', () => {
    render(<ShopItem onClick={() => {}} />)
    const showBtn = screen.getByRole('button', { name: 'Show Detail'})
    const backBtn = screen.getByRole('button', { hidden: true, name: 'Back to Shop' })
    expect(showBtn).toBeInTheDocument()
    expect(backBtn).toBeInTheDocument()
  })

  it('renders all buttons with the correct text', () => {
    render(<ShopItem />)
    expect(screen.getAllByRole('button')[0].textContent).toMatch('Show Detail')
    expect(screen.getAllByRole('button', { hidden: true })[1].textContent).toMatch('Back to Shop')
  })

  it('renders the DetailItem dialog when button clicked', async () => {
    HTMLDialogElement.prototype.showModal = vi.fn()
    HTMLDialogElement.prototype.close = vi.fn()
    render(<ShopItem />)
    const user = userEvent.setup()
    const showBtn = screen.getByRole('button', { name: 'Show Detail'})
    await user.click(showBtn)
    expect(HTMLDialogElement.prototype.showModal).toHaveBeenCalled()
  })

  it('closes the DetailItem dialog when button clicked', async () => {
    HTMLDialogElement.prototype.showModal = vi.fn()
    HTMLDialogElement.prototype.close = vi.fn()
    render(<ShopItem />)
    const user = userEvent.setup()
    const showBtn = screen.getByRole('button', { name: 'Show Detail'})
    await user.click(showBtn)
    expect(HTMLDialogElement.prototype.showModal).toHaveBeenCalled()
    const backBtn = screen.getByRole('button', { hidden: true, name: 'Back to Shop' })
    await user.click(backBtn)
    expect(HTMLDialogElement.prototype.close).toHaveBeenCalled()
  })
})
