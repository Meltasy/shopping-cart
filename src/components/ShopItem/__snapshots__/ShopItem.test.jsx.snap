// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop item component > renders shop product item 1`] = `
<div>
  <div
    class="sc-jBIHhB loGoFA"
  >
    <img
      alt=" "
      class="sc-cMqPwU duECSg"
    />
    <div
      class="sc-jUkaYT bHrHaU"
    >
      <h3 />
      <div
        class="sc-jHswkR eqKxLb"
      >
        <div
          class="sc-deSjom hTmSZG"
        >
          <h3>
            €
            0.00
          </h3>
          <button
            class="sc-cSaEAk dpGhXn"
          >
            <svg
              role="presentation"
              style="width: 1.5rem; height: 1.5rem;"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5,14L20.5,19L19,20.5L14,15.5V14.71L13.73,14.43C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.43,13.73L14.71,14H15.5M9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14M12,10H10V12H9V10H7V9H9V7H10V9H12V10Z"
                style="fill: currentColor;"
              />
            </svg>
          </button>
          <dialog
            class="sc-FEMpB hjZXnf"
          >
            <div
              class="sc-gjcoXW hKdEaB"
            >
              <div
                class="sc-dVBluf hrHbBw"
              >
                <div
                  class="sc-fLDLck lhfGYe"
                >
                  <h2 />
                  <h2>
                    €
                    0.00
                  </h2>
                </div>
                <div>
                  <button
                    class="sc-cSaEAk dpGhXn"
                  >
                    <svg
                      role="presentation"
                      style="width: 2.25rem; height: 2.25rem;"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M15.5,14H14.71L14.43,13.73C15.41,12.59 16,11.11 16,9.5A6.5,6.5 0 0,0 9.5,3A6.5,6.5 0 0,0 3,9.5A6.5,6.5 0 0,0 9.5,16C11.11,16 12.59,15.41 13.73,14.43L14,14.71V15.5L19,20.5L20.5,19L15.5,14M9.5,14C7,14 5,12 5,9.5C5,7 7,5 9.5,5C12,5 14,7 14,9.5C14,12 12,14 9.5,14M7,9H12V10H7V9Z"
                        style="fill: currentColor;"
                      />
                    </svg>
                  </button>
                </div>
              </div>
              <img
                alt=" "
                class="sc-dycHyt hSmkZk"
              />
              <p />
            </div>
          </dialog>
        </div>
        <form
          class="sc-jbAkgO knpBS"
        >
          <div>
            <button
              class="sc-gNZgCX iwZfDw"
              type="button"
            >
              -
            </button>
            <input
              class="sc-gNZgCX sc-hRDKVd iwZfDw eYnWSA"
              id="quantity"
              max="10"
              min="1"
              name="quantity"
              pattern="^\\d+$"
              type="text"
              value="1"
            />
            <button
              class="sc-gNZgCX iwZfDw"
              type="button"
            >
              +
            </button>
          </div>
          <button
            class="sc-kfeOyU cvoqkg"
            type="submit"
          >
            <svg
              role="presentation"
              style="width: 1.5rem; height: 1.5rem;"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 15V18H23V20H20V23H18V20H15V18H18V15H20M12 13C10.9 13 10 13.9 10 15S10.9 17 12 17 14 16.1 14 15 13.1 13 12 13M13.35 21H5.5C4.58 21 3.81 20.38 3.58 19.54L1.04 10.27C1 10.18 1 10.09 1 10C1 9.45 1.45 9 2 9H6.79L11.17 2.45C11.36 2.16 11.68 2 12 2S12.64 2.16 12.83 2.44L17.21 9H22C22.55 9 23 9.45 23 10L22.97 10.27L22 13.81C21.43 13.5 20.79 13.24 20.12 13.11L20.7 11H3.31L5.5 19H13C13 19.7 13.13 20.37 13.35 21M9.2 9H14.8L12 4.8L9.2 9Z"
                style="fill: currentColor;"
              />
            </svg>
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
